### DISCUSSION TITLE
How to specify EnvironmentName when deploying to Azure Container App?

### DISCUSSION THREAD

**Discussion by RamType0** [microsoft/aspire#5797] [Q&A]
How to specify EnvironmentName when deploying to Azure Container App?
When we are running Aspire on local computer, we can specify it in launchsettings.json.
Is there something like publishProfile for Aspire AppHost?

**Comment by behdad088** [comment]
Hi @RamType0 

When you're running your app locally, you can easily set the environment (like `Development`, `Production`, etc.) using `launchSettings.json`. But when you're deploying to **Azure Container Apps**, there's no `launchSettings.json` anymore. Instead, you need to set it differently, like using environment variables.

There are multiple ways to do it:

### 1. **Azure Portal (Easy Way)**
You can set the environment variable directly in the Azure Portal:
- Go to your **Azure Container App**.
- In **Settings**, click **Containers**.
- Under **Environment Variables**, add a new one:
  - Name: `ASPNETCORE_ENVIRONMENT`
  - Value: `Production` (or `Development`, whatever you need).
  
Boom! Now Azure knows which environment to run in.

### 2. **Azure CLI (Command Line)**
If you're more of a command line person or you're automating deployments, you can set the environment variable with Azure CLI.

Just run this command:

```bash
az containerapp update --name <app-name> \
       --resource-group <resource-group> \
       --set-env-vars "ASPNETCORE_ENVIRONMENT=Production"
```

This tells Azure to run the app in "Production" mode.

### 3. **Dockerfile**
If you're using Docker to build your app, you can include the environment variable right in your `Dockerfile`. It’s simple, just add:

```dockerfile
ENV ASPNETCORE_ENVIRONMENT=Production
```

This way, when you build your container and deploy it, Azure will know to run it in `Production` mode.

### 4. **Bicep/ARM Templates (For Infra Nerds)**
If you’re deploying using infrastructure-as-code (like Bicep or ARM templates), you can define the environment variable in the template itself. Here’s a quick example for Bicep:

```bicep
resource containerApp 'Microsoft.Web/containerApps@2022-01-01-preview' = {
  properties: {
    configuration: {
      environmentVariables: [
        {
          name: 'ASPNETCORE_ENVIRONMENT'
          value: 'Production'
        }
      ]
    }
  }
}
```

This makes sure the app runs in `Production`.

### 5. **GitHub Actions (CI/CD Pipelines)**
If you're using GitHub Actions for deploying, you can add the environment variable in your workflow YAML file. Here’s how:

```yaml
jobs:
  build-and-deploy:
    steps:
    - name: Deploy to Azure
      run: |
        az containerapp update --name <app-name> \
          --resource-group <resource-group> \
          --set-env-vars "ASPNETCORE_ENVIRONMENT=Production"
```

### So Basically:
No more `launchSettings.json` on Azure. You just set the `ASPNETCORE_ENVIRONMENT` environment variable wherever it fits (Portal, CLI, Docker, etc.), and that’s it. Azure will know how to run the app based on the value you give it, like `Production` or `Development`. Easy!

**Reply by davidfowl** [reply]
You missed the azd way in your instructions.

For this though you would just use WithEnvironment

**Reply by RamType0** [reply]
How can we switch environment name(e.g. Production, Development) with the solution using `WithEnvironement`?

**Reply by davidfowl** [reply]
Can you show me your apphost code?

**Reply by RamType0** [reply]
```C#
using Microsoft.Extensions.Hosting;

var builder = DistributedApplication.CreateBuilder(args);

// "Development" means local development. I'm using "Staging" or "Production" for Azure Container Apps.

var globalConfiguration = builder.Environment.IsDevelopment() && builder.ExecutionContext.IsRunMode
    ? builder.AddConnectionString("global-configuration")
    : builder.AddAzureAppConfiguration("global-configuration");


const string applicationDatabaseSqlDatabaseName = "application_db";
var postgres = builder.AddPostgres("postgres").PublishAsAzurePostgresFlexibleServer()
    .WithEnvironment("POSTGRES_DB", applicationDatabaseSqlDatabaseName)
    .WithDataVolume("postgres.data")
    .WithPgAdmin(containerName: "pg-admin");

var applicationDb = postgres.AddDatabase("application-db", applicationDatabaseSqlDatabaseName);


var migrationService = builder.AddProject<Projects.Data_MigrationService>("migration-service")
    .WithReference(applicationDb)
    .WithReference(globalConfiguration);

var storage = builder.AddAzureStorage("storage");

if (builder.Environment.IsDevelopment())
{
    storage.RunAsEmulator(options => options.WithDataVolume());
}

var blobs = storage.AddBlobs("blobs");


var cache = builder.AddRedis("cache");


var azureSignalR = builder.Environment.IsDevelopment() && builder.ExecutionContext.IsRunMode
    ? builder.AddConnectionString("azure-signal-r")
    : builder.AddAzureSignalR("azure-signal-r");

var apiServiceSecrets = builder.Environment.IsDevelopment() && builder.ExecutionContext.IsRunMode
    ? builder.AddConnectionString("api-service-secrets") 
    : builder.AddAzureKeyVault("api-service-secrets");

var apiService = builder.AddProject<Projects.ApiService>("api-service")
    .WithReference(applicationDb)
    .WithReference(blobs)
    .WithReference(azureSignalR)
    .WithReference(apiServiceSecrets)
    .WithReference(globalConfiguration);


var webFrontend = builder.AddProject<Projects.Web>("web-frontend")
    .WithExternalHttpEndpoints()
    .WithReference(cache)
    .WithReference(apiService)
    .WithReference(globalConfiguration);

apiService.WithEnvironment("WebFrontendUrl", webFrontend.GetEndpoint("https"));
webFrontend.WithEnvironment("ApiEndpoint", apiService.GetEndpoint("https"));

// https://github.com/dotnet/aspire/issues/3145#issuecomment-2241799215
if (builder.Environment.IsDevelopment() && builder.ExecutionContext.IsRunMode)
{
    webFrontend.WithEndpoint("https", endpoint => endpoint.IsProxied = false);
}
var managementConsoleSecrets = builder.Environment.IsDevelopment() && builder.ExecutionContext.IsRunMode
    ? builder.AddConnectionString("manager-secrets")
    : builder.AddAzureKeyVault("manager-secrets");

var managementConsole = builder.AddProject<Projects.ManagementConsole>("management-console")
    .WithExternalHttpEndpoints()
    .WithReference(applicationDb)
    .WithReference(blobs)
    .WithReference(managementConsoleSecrets)
    .WithReference(globalConfiguration);




builder.Build().Run();

```

**Comment by RamType0** [comment]
Azure Portal Environment variable settings will be reset every time we deploy, isn't it?

**Reply by behdad088** [reply]
It really depends on how you deploy. For example you f you use ci cd you can set the environment variables. If you also directly deploy to the app services does not override the configuration

**Reply by RamType0** [reply]
> It really depends on how you deploy

When I deploy with visual studio GUI, then set environment variables from Azure portal, then deploy with visual studio GUI again,
Environment variable settings will be reset. Isn't it?

**Reply by behdad088** [reply]
No it will not reset the environment variable in app service

**Reply by RamType0** [reply]
I just tried it by myself, but environment variable has been reset after I re-deploying it.

**Reply by davidfowl** [reply]
It will get reset for sure. When you deploy the app model is the source of truth.

**Comment by argium** [comment]
I've added a custom domain to my application and I deploy using `azd up`. I need to update the value of the environment variable `services__apiservice__https__0` so that it doesn't get overwritten on subsequent deployments but I'm having difficulty finding where the right place is. The documentation only covers launch settings.json for local development. I could use appsettings.json but the container environment variables will probably take precedence.
